File: ga/gpt3_5_result_2024_6/result/tree_gpt/11_tree.c
use std::ptr;

type ElementType = i32;

struct TreeNode {
    element: ElementType,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

type SearchTree = Option<Box<TreeNode>>;
type Position = Option<Box<TreeNode>>;

fn make_empty(t: SearchTree) -> SearchTree {
    match t {
        Some(mut node) => {
            node.left = make_empty(node.left);
            node.right = make_empty(node.right);
            None
        }
        None => None,
    }
}

fn find(x: ElementType, t: SearchTree) -> Position {
    match t {
        Some(node) => {
            if x < node.element {
                find(x, node.left.clone())
            } else if x > node.element {
                find(x, node.right.clone())
            } else {
                Some(node)
            }
        }
        None => None,
    }
}

fn find_min(t: SearchTree) -> Position {
    match t {
        Some(mut node) => {
            if let Some(mut left) = node.left {
                find_min(Some(left))
            } else {
                Some(node)
            }
        }
        None => None,
    }
}

fn find_max(t: SearchTree) -> Position {
    match t {
        Some(mut node) => {
            while let Some(mut right) = node.right {
                node = right;
            }
            Some(node)
        }
        None => None,
    }
}

fn insert(x: ElementType, t: SearchTree) -> SearchTree {
    match t {
        Some(mut node) => {
            if x < node.element {
                node.left = insert(x, node.left);
            } else if x > node.element {
                node.right = insert(x, node.right);
            }
            Some(node)
        }
        None => {
            let new_node = Box::new(TreeNode {
                element: x,
                left: None,
                right: None,
            });
            Some(new_node)
        }
    }
}

fn delete(x: ElementType, t: SearchTree) -> SearchTree {
    match t {
        Some(mut node) => {
            if x < node.element {
                node.left = delete(x, node.left);
                Some(node)
            } else if x > node.element {
                node.right = delete(x, node.right);
                Some(node)
            } else {
                if let (Some(mut left), Some(right)) = (node.left, node.right) {
                    let min_node = find_min(Some(right));
                    let min_element = min_node.as_ref().unwrap().element;
                    node.element = min_element;
                    node.right = delete(min_element, node.right);
                    Some(node)
                } else {
                    let temp = node;
                    if node.left.is_none() {
                        return node.right;
                    } else if node.right.is_none() {
                        return node.left;
                    }
                    drop(temp);
                    None
                }
            }
        }
        None => None,
    }
}

fn retrieve(p: Position) -> ElementType {
    p.unwrap().element
}

fn main() {
    let mut t: SearchTree = None;
    let mut j = 0;
  
    t = make_empty(t);
    for i in 0..50 {
        j = (j + 7) % 50;
        t = insert(j, t);
    }
    for i in 0..50 {
        let p = find(i, t.clone());
        if p.is_none() || retrieve(p) != i {
            println!("Error at {}", i);
        }
    }

    for i in (0..50).step_by(2) {
        t = delete(i, t);
    }

    for i in (1..50).step_by(2) {
        let p = find(i, t.clone());
        if p.is_none() || retrieve(p) != i {
            println!("Error at {}", i);
        }
    }
    for i in (0..50).step_by(2) {
        let p = find(i, t.clone());
        if !p.is_none() {
            println!("Error at {}", i);
        }
    }

    println!("Min is {}, Max is {}", retrieve(find_min(t).unwrap()), retrieve(find_max(t).unwrap()));
}

File: ga/glm4_result_2024_6/result/tree_glm/11_tree.rs
use std::cmp::Ordering;
use std::fmt;

#[derive(Debug)]
struct TreeNode {
    element: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

impl TreeNode {
    fn new(element: i32) -> Self {
        TreeNode {
            element,
            left: None,
            right: None,
        }
    }
}

impl fmt::Display for TreeNode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({})", self.element)
    }
}

#[derive(Debug)]
struct SearchTree(Option<Box<TreeNode>>);

impl SearchTree {
    fn make_empty(&mut self) {
        self.0.take();
    }

    fn find(&self, x: i32) -> Option<&TreeNode> {
        let mut current = self.0.as_ref();
        while let Some(node) = current {
            match x.cmp(&node.element) {
                Ordering::Less => current = node.left.as_ref(),
                Ordering::Greater => current = node.right.as_ref(),
                Ordering::Equal => return Some(node),
            }
        }
        None
    }

    fn find_min(&self) -> Option<&TreeNode> {
        let mut current = self.0.as_ref();
        while let Some(node) = current {
            if node.left.is_none() {
                return Some(node);
            }
            current = node.left.as_ref();
        }
        None
    }

    fn find_max(&self) -> Option<&TreeNode> {
        let mut current = self.0.as_ref();
        while let Some(node) = current {
            if node.right.is_none() {
                return Some(node);
            }
            current = node.right.as_ref();
        }
        None
    }

    fn insert(&mut self, x: i32) {
        match self.0.as_mut() {
            None => self.0 = Some(Box::new(TreeNode::new(x))),
            Some(node) => match x.cmp(&node.element) {
                Ordering::Less => {
                    if node.left.is_none() {
                        node.left = Some(Box::new(TreeNode::new(x)));
                    } else {
                        self.insert(x);
                    }
                }
                Ordering::Greater => {
                    if node.right.is_none() {
                        node.right = Some(Box::new(TreeNode::new(x)));
                    } else {
                        self.insert(x);
                    }
                }
                Ordering::Equal => {}
            },
        }
    }

    fn delete(&mut self, x: i32) {
        if let Some(mut node) = self.0.take() {
            match x.cmp(&node.element) {
                Ordering::Less => {
                    if let Some(left) = node.left.take() {
                        self.0 = Some(left);
                        self.delete(x);
                    }
                }
                Ordering::Greater => {
                    if let Some(right) = node.right.take() {
                        self.0 = Some(right);
                        self.delete(x);
                    }
                }
                Ordering::Equal => {
                    if let (Some(left), Some(right)) = (node.left.take(), node.right.take()) {
                        self.0 = Some(left);
                        self.insert(right.element);
                    } else if let Some(left) = node.left.take() {
                        self.0 = Some(left);
                    } else if let Some(right) = node.right.take() {
                        self.0 = Some(right);
                    } else {
                        self.0 = None;
                    }
                }
            }
        }
    }

    fn retrieve(&self) -> Option<i32> {
        self.0.as_ref().map(|node| node.element)
    }
}

fn main() {
    let mut tree = SearchTree(None);
    for i in 0..50 {
        tree.insert(i);
    }
    for i in 0..50 {
        assert_eq!(tree.find(i), Some(&TreeNode::new(i)));
    }
    for i in 0..50 {
        if i % 2 == 0 {
            tree.delete(i);
        }
    }
    for i in 0..50 {
        if i % 2 == 1 {
            assert_eq!(tree.find(i), Some(&TreeNode::new(i)));
        } else {
            assert_eq!(tree.find(i), None);
        }
    }
    println!("Min is {}, Max is {}", tree.find_min().unwrap(), tree.find_max().unwrap());
}